File: freestylemasonry/submissions/accepted/jannik.hs
import qualified Data.ByteString.Char8 as C
import Data.List
import Data.Bool

int s = case C.readInt s of Just (x,_) -> x
main = C.getContents >>= putStrLn . solve . map int . C.words

solve :: [Int] -> String
solve (_:h:hs) = bool "impossible" "possible" possible
  where
    possible = h == foldl' (\top x -> bool
        (-1)
        (min h $ top + 1 - 2 * mod (top - x) 2)
        (top >= x)) h hs

File: freestylemasonry/submissions/accepted/jannik.java
import java.util.*;

public class jannik {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int w = in.nextInt(), h = in.nextInt();
		int top = h;
		while (w-- > 0 && top >= 0) {
			int x = in.nextInt();
			if (x > top) top = -5;
			top = Math.min(h, top + 1 - 2 * ((top - x)%2));
		}
		System.out.println(top == h ? "possible" : "impossible");
	}
}
